modules/@apostrophecms/log/index.js

```diff
@@ -0,0 +1,255 @@
+      // the message.
+      // `data.severity` is always set to the value of the `severity` argument.
+      // `data.type` is always set to the value of the `eventType` argument.
+      // message is optional, if not provided it is generated from the eventType and data.module.
+      // Optional message values:
+      // - `module: eventType: message`
+      // - `eventType: message`: (missing module)
+      // - `module: eventType`: (missing message)
+      // - `eventType`: (missing module and message)
+      //
+      // Returns [ data ] or [ message, data ] depending on option.messageAs.
+      // `data` is always an object containing at least a `type` and `severity` properties.
+      processLoggerArgs(...allArgs) {
+        const args = allArgs.slice(1);
+        const moduleSelf = allArgs[0];
+        let severity;
+        let req;
+        let eventType;
+        let message;
+        let data;
+
+        // Detect `req` argument with a simple duck type check - apos `req` object
+        // has always a translate `t` function.
+        if (args[1] && typeof args[1].t === 'function') {
+          [ severity, req, eventType, message, data ] = args;
+        } else {
+          [ severity, eventType, message, data ] = args;
+        }
+
+        if (!_.isString(eventType)) {
+          throw new Error('Event type must be a string');
+        }
+        if (_.isPlainObject(message)) {
+          data = message;
+          message = undefined;
+        }
+        data = data ? { ...data } : {};
+        data.type = eventType;
+        data.severity = severity;
+        if (moduleSelf) {
+          data.module = moduleSelf.__meta?.name;
+        }
+
+        if (_.isString(message) && message.trim().length > 0) {
+          message = data.module
+            ? `${data.module}: ${eventType}: ${message}`
+            : `${eventType}: ${message}`;
+        } else {
+          message = data.module
+            ? `${data.module}: ${eventType}`
+            : eventType;
+        }
+
+        data = self.processRequestData(req, data);
+
+        if (self.options.messageAs) {
+          data[self.options.messageAs] = message;
+          return [ data ];
+        }
+
+        return [ message, data ];
+      },
+
+      // Enrich the `data` argument with additional information from the request.
+      processRequestData(req, data) {
+        if (!req) {
+          return data;
+        }
+        return {
+          // https://expressjs.com/en/api.html#req.originalUrl
+          url: req.originalUrl,
+          path: req.path,
+          ip: req.ip,
+          query: req.query,
+          requestId: req.requestId || self.apos.util.generateId(),
+          ...data
+        };
+      },
+
+      // Assess the module filter configuration and determine if the log
+      // should be kept or rejected.
+      //
+      // `data` argument should be an object returned from `processLogArgs(...)`.
+      shouldKeepEntry(data) {
```

Comment:
Try to do this really early. It would be great if we didn't have to clone any objects before this happens.

Reply:
Make sense. The problem is that I need the `data` object already formatted so I can assess it. But I'll look into it closely today.